src/main/java/io/github/fablabsmc/fablabs/api/fluidvolume/v1/math/Fraction.java

```diff
@@ -209,55 +140,172 @@ public static Fraction min(Fraction left, Fraction right) {
+		int[] arr = dynamic.asIntStream().toArray();
+
+		if (arr.length == 0) {
+			return ZERO;
+		}
+
+		if (arr.length == 1) {
+			return ofWhole(arr[0]);
+		}
+
+		return of(arr[0], arr[1]);
+	}
+
+	public int getNumerator(int denominator) {
+		return this.numerator * (denominator / this.denominator);
+	}
+
+	public int getNumerator() {
+		return this.numerator;
+	}
+
+	public boolean divisible(Fraction fraction) {
+		return this.divide(fraction).getDenominator() == 1;
+	}
+
+	public int getDenominator() {
+		return this.denominator;
+	}
+
+	public Fraction divide(Fraction other) {
+		return this.multiply(other.inverse());
+	}
+
+	public Fraction multiply(Fraction other) {
+		int gcd1 = IntMath.gcd(Math.abs(this.numerator), other.denominator);
+		int gcd2 = IntMath.gcd(this.denominator, Math.abs(other.numerator));
+		// guaranteed simplified
+		return new Fraction(this.signum() * other.signum() * (this.numerator / gcd1) * (other.numerator / gcd2), (this.denominator / gcd2) * (other.denominator / gcd1));
+	}
+
+	public Fraction inverse() throws ArithmeticException {
+		// don't need to simplify
+		switch (this.signum()) {
+		case 1:
+			return new Fraction(this.denominator, this.numerator);
+		case -1:
+			return new Fraction(-this.denominator, -this.numerator);
+		default:
+			throw new ArithmeticException("Cannot invert zero fraction!");
+		}
+	}
+
+	public int signum() {
+		return Integer.signum(this.numerator);
+	}
+
+	public Fraction floorNearest(Fraction fraction) {
+		return fraction.multiply(ofWhole(this.floorDiv(fraction)));
+	}
+
+	public int floorDiv(Fraction fraction) {
+		return Math.floorDiv(this.numerator * fraction.denominator, fraction.numerator * this.denominator);
+	}
+
+	public boolean isNegative() {
+		return this.numerator < 0;
+	}
+
+	public boolean isPositive() {
+		return this.numerator > 0;
+	}
+
 	@Override
-	public int intValue() {
-		return numerator / denominator;
+	public double doubleValue() {
+		return ((double) this.numerator) / this.denominator;
+	}
+
+	public Fraction negate() {
+		// don't need to simplify
+		return new Fraction(-this.numerator, this.denominator);
+	}
+
+	public Fraction multiply(int amount) {
+		return of(this.numerator * amount, this.denominator);
+	}
+
+	public Fraction add(Fraction other) {
+		int commonMultiple = lcm(this.denominator, other.denominator);
+		int leftNumerator = commonMultiple / this.denominator * this.numerator;
+		int rightNumerator = commonMultiple / other.denominator * other.numerator;
+		return ofValidDenominator(leftNumerator + rightNumerator, commonMultiple);
+	}
+
+	public Fraction subtract(Fraction other) {
+		int commonMultiple = lcm(this.denominator, other.denominator);
+		int leftNumerator = commonMultiple / this.denominator * this.numerator;
+		int rightNumerator = commonMultiple / other.denominator * other.numerator;
+		return ofValidDenominator(leftNumerator - rightNumerator, commonMultiple);
+	}
+
+	public Fraction floorWithDenominator(int denom) {
```

Comment:
```suggestion
	public Fraction floorWithDenominator(int denominator) {
```

We really should be using full names for `denominator`, since `denom` doesn't mean much to someone glancing from far.